File: clients/go/nubdb.go
// Package nubdb provides a Go client for NubDB database.
package nubdb

import (
"bufio"
"fmt"
"net"
"strconv"
"strings"
"time"
)

// Client represents a connection to NubDB
type Client struct {
conn   net.Conn
reader *bufio.Reader
writer *bufio.Writer
host   string
port   int
}

// Config holds configuration for the client
type Config struct {
Host    string
Port    int
Timeout time.Duration
}

// DefaultConfig returns default configuration
func DefaultConfig() *Config {
return &Config{
Host:    "localhost",
Port:    6379,
Timeout: 5 * time.Second,
}
}

// Connect creates a new connection to NubDB
func Connect(config *Config) (*Client, error) {
if config == nil {
config = DefaultConfig()
}

addr := fmt.Sprintf("%s:%d", config.Host, config.Port)
conn, err := net.DialTimeout("tcp", addr, config.Timeout)
if err != nil {
return nil, fmt.Errorf("failed to connect: %w", err)
}

client := &Client{
conn:   conn,
reader: bufio.NewReader(conn),
writer: bufio.NewWriter(conn),
host:   config.Host,
port:   config.Port,
}

return client, nil
}

// sendCommand sends a command and returns the response
func (c *Client) sendCommand(cmd string) (string, error) {
// Write command
_, err := c.writer.WriteString(cmd + "\n")
if err != nil {
return "", fmt.Errorf("write error: %w", err)
}

if err := c.writer.Flush(); err != nil {
return "", fmt.Errorf("flush error: %w", err)
}

// Read response
response, err := c.reader.ReadString('\n')
if err != nil {
return "", fmt.Errorf("read error: %w", err)
}

return strings.TrimSpace(response), nil
}

// Set stores a key-value pair
func (c *Client) Set(key, value string, ttl int) error {
cmd := fmt.Sprintf(`SET %s "%s"`, key, value)
if ttl > 0 {
cmd += fmt.Sprintf(" %d", ttl)
}

response, err := c.sendCommand(cmd)
if err != nil {
return err
}

if response != "OK" {
return fmt.Errorf("unexpected response: %s", response)
}

return nil
}

// Get retrieves a value by key
func (c *Client) Get(key string) (string, error) {
response, err := c.sendCommand(fmt.Sprintf("GET %s", key))
if err != nil {
return "", err
}

if response == "(nil)" {
return "", nil
}

// Remove quotes if present
response = strings.Trim(response, `"`)
return response, nil
}

// Delete removes a key
func (c *Client) Delete(key string) error {
response, err := c.sendCommand(fmt.Sprintf("DELETE %s", key))
if err != nil {
return err
}

if response != "OK" && response != "(not found)" {
return fmt.Errorf("unexpected response: %s", response)
}

return nil
}

// Exists checks if a key exists
func (c *Client) Exists(key string) (bool, error) {
response, err := c.sendCommand(fmt.Sprintf("EXISTS %s", key))
if err != nil {
return false, err
}

return response == "1", nil
}

// Incr increments a counter
func (c *Client) Incr(key string) (int64, error) {
response, err := c.sendCommand(fmt.Sprintf("INCR %s", key))
if err != nil {
return 0, err
}

value, err := strconv.ParseInt(response, 10, 64)
if err != nil {
return 0, fmt.Errorf("invalid response: %s", response)
}

return value, nil
}

// Decr decrements a counter
func (c *Client) Decr(key string) (int64, error) {
response, err := c.sendCommand(fmt.Sprintf("DECR %s", key))
if err != nil {
return 0, err
}

value, err := strconv.ParseInt(response, 10, 64)
if err != nil {
return 0, fmt.Errorf("invalid response: %s", response)
}

return value, nil
}

// Size returns the number of keys
func (c *Client) Size() (int64, error) {
response, err := c.sendCommand("SIZE")
if err != nil {
return 0, err
}

// Parse "N keys" format
parts := strings.Fields(response)
if len(parts) > 0 {
value, err := strconv.ParseInt(parts[0], 10, 64)
if err != nil {
return 0, fmt.Errorf("invalid response: %s", response)
}
return value, nil
}

return 0, fmt.Errorf("invalid response: %s", response)
}

// Clear deletes all keys
func (c *Client) Clear() error {
response, err := c.sendCommand("CLEAR")
if err != nil {
return err
}

if response != "OK" {
return fmt.Errorf("unexpected response: %s", response)
}

return nil
}

// Close closes the connection
func (c *Client) Close() error {
if c.conn != nil {
c.sendCommand("QUIT")
return c.conn.Close()
}
return nil
}

File: clients/rust/nubdb.rs
//! NubDB Rust Client
//! 
//! Simple client library for connecting to NubDB database.

use std::io::{BufRead, BufReader, Write};
use std::net::TcpStream;

pub struct NubDB {
    stream: TcpStream,
    reader: BufReader<TcpStream>,
}

impl NubDB {
    /// Connect to NubDB server
    pub fn connect(addr: &str) -> Result<Self, std::io::Error> {
        let stream = TcpStream::connect(addr)?;
        let reader = BufReader::new(stream.try_clone()?);
        
        Ok(NubDB { stream, reader })
    }

    /// Send a command and get response
    fn send_command(&mut self, cmd: &str) -> Result<String, std::io::Error> {
        writeln!(self.stream, "{}", cmd)?;
        self.stream.flush()?;

        let mut response = String::new();
        self.reader.read_line(&mut response)?;
        
        Ok(response.trim().to_string())
    }

    /// SET key-value pair
    pub fn set(&mut self, key: &str, value: &str, ttl: Option<u32>) -> Result<bool, std::io::Error> {
        let cmd = match ttl {
            Some(t) => format!(r#"SET {} "{}" {}"#, key, value, t),
            None => format!(r#"SET {} "{}""#, key, value),
        };

        let response = self.send_command(&cmd)?;
        Ok(response == "OK")
    }

    /// GET value by key
    pub fn get(&mut self, key: &str) -> Result<Option<String>, std::io::Error> {
        let response = self.send_command(&format!("GET {}", key))?;
        
        if response == "(nil)" {
            Ok(None)
        } else {
            // Remove quotes
            let value = response.trim_matches('"').to_string();
            Ok(Some(value))
        }
    }

    /// DELETE key
    pub fn delete(&mut self, key: &str) -> Result<bool, std::io::Error> {
        let response = self.send_command(&format!("DELETE {}", key))?;
        Ok(response == "OK")
    }

    /// EXISTS check if key exists
    pub fn exists(&mut self, key: &str) -> Result<bool, std::io::Error> {
        let response = self.send_command(&format!("EXISTS {}", key))?;
        Ok(response == "1")
    }

    /// INCR increment counter
    pub fn incr(&mut self, key: &str) -> Result<i64, std::io::Error> {
        let response = self.send_command(&format!("INCR {}", key))?;
        response.parse::<i64>()
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))
    }

    /// DECR decrement counter
    pub fn decr(&mut self, key: &str) -> Result<i64, std::io::Error> {
        let response = self.send_command(&format!("DECR {}", key))?;
        response.parse::<i64>()
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))
    }

    /// SIZE get number of keys
    pub fn size(&mut self) -> Result<usize, std::io::Error> {
        let response = self.send_command("SIZE")?;
        let parts: Vec<&str> = response.split_whitespace().collect();
        
        if let Some(num_str) = parts.first() {
            num_str.parse::<usize>()
                .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))
        } else {
            Ok(0)
        }
    }

    /// CLEAR delete all keys
    pub fn clear(&mut self) -> Result<bool, std::io::Error> {
        let response = self.send_command("CLEAR")?;
        Ok(response == "OK")
    }

    /// Close connection
    pub fn close(&mut self) -> Result<(), std::io::Error> {
        self.send_command("QUIT")?;
        Ok(())
    }
}

// Example usage
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_operations() {
        let mut client = NubDB::connect("localhost:6379").unwrap();
        
        // SET
        assert!(client.set("name", "Alice", None).unwrap());
        
        // GET
        let value = client.get("name").unwrap();
        assert_eq!(value, Some("Alice".to_string()));
        
        // EXISTS
        assert!(client.exists("name").unwrap());
        
        // DELETE
        assert!(client.delete("name").unwrap());
        assert!(!client.exists("name").unwrap());
        
        client.close().unwrap();
    }
}
